libpromises/override_fsattrs.c

```diff
@@ -0,0 +1,140 @@
+#include <override_fsattrs.h>
+#include <platform.h>
+#include <fsattrs.h>
+#include <logging.h>
+#include <stdlib.h>
+#include <files_copy.h>
+#include <cf3.defs.h>
+#include <string_lib.h>
+
+bool OverrideImmutableBegin(const char *orig, char *copy, size_t copy_len)
+{
+    srand(time(NULL)); /* Seed random number generator */
+    int rand_number = rand() % 999999;
+    assert(rand_number >= 0);
+
+    /* Inspired by mkstemp(3) */
+    int ret = snprintf(copy, copy_len, "%s.%06d" CF_NEW, orig, rand_number);
+    if (ret < 0 || (size_t) ret >= copy_len)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to generate name for temporary copy of '%s': Filename is too long (%d >= %zu)",
+            orig,
+            ret,
+            copy_len);
+        return false;
+    }
+
+    /* We'll match the original file permissions on commit */
+    if (!CopyRegularFileDiskPerms(orig, copy, 0600))
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to copy file '%s' to temporary file '%s'",
+            orig,
+            copy);
+        return false;
+    }
+
+    return true;
+}
+
+bool OverrideImmutableCommit(const char *orig, const char *copy)
+{
+    struct stat sb;
+    if (lstat(orig, &sb) == -1)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to stat file '%s'", orig);
+        unlink(copy);
+        return false;
+    }
+
+    if (chmod(copy, sb.st_mode) == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to change mode bits on file '%s' to %04jo: %s",
```

Comment:
same here, seems like adding a bit of context to the message would be helpful when debugging.